smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,390 @@
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.utils.MapUtils;
+import software.amazon.smithy.utils.SmithyBuilder;
+import software.amazon.smithy.utils.ToSmithyBuilder;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.io.Writer;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Class that represents the contents of a Smithy trace file.
+ * TraceFile's require a smithyTrace file version number, {@link ArtifactMetadata}, and
+ * {@link Map} from {@link ShapeId} to a List of {@link ShapeLink} objects. TraceFile's
+ * optionally have a {@link ArtifactDefinitions} object.
+ * <p>
+ * TraceFile handles parsing, serialization and deserialization of a Smithy trace file.
+ * </p>
+ * <p>
+ * The smithyTrace {@link String} contains the Smithy trace file version number.
+ * </p>
+ * <p>
+ * The shapes {@link Map} provides a mapping of absolute Smithy shape IDs to a list
+ * of shape link objects. A single Smithy shape can be responsible for generating
+ * multiple components in the target artifact.
+ * </p>
+ */
+public final class TraceFile implements ToNode, ToSmithyBuilder<TraceFile> {
+    public static final String SMITHY_TRACE_TEXT = "smithyTrace";
+    public static final String ARTIFACT_TEXT = "artifact";
+    public static final String DEFINITIONS_TEXT = "definitions";
+    public static final String SHAPES_TEXT = "shapes";
+    public static final String SMITHY_TRACE_VERSION = "1.0";
+
+    private String smithyTrace;
+    private ArtifactMetadata artifactMetadata;
+    private ArtifactDefinitions artifactDefinitions; //Optional
+    private Map<ShapeId, List<ShapeLink>> shapes;
+    private SourceLocation sl = new SourceLocation("");
+
+    private TraceFile(Builder builder) {
+        smithyTrace = SmithyBuilder.requiredState(SMITHY_TRACE_TEXT, builder.smithyTrace);
+        artifactMetadata = SmithyBuilder.requiredState(ARTIFACT_TEXT, builder.artifactMetadata);
+        shapes = SmithyBuilder.requiredState(SHAPES_TEXT, MapUtils.copyOf(builder.shapes));
+        artifactDefinitions = builder.artifactDefinitions;
+    }
+
+    /**
+     * Parses and validates the trace file passed in as filename
+     * and instantiates smithyTrace and defintions, and fills
+     * artifactMetadata and shapes.
+     *
+     * @param filename the absolute or relative path of tracefile
+     * @throws FileNotFoundException if filename is not found for reading
+     */
+    public static TraceFile parseTraceFile(URI filename) throws FileNotFoundException {
```

Comment:
Probably should remove this, see the previous comment about separation of concerns with file loading.